seq2seq/adapters/adapter_controller.py

```diff
@@ -0,0 +1,193 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implements Adapter Controller, a module that keeps multiple 
+layers of Adapters, and controls which adapter layer to use."""
+import os
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+from .adapter_configuration import AdapterConfig, MetaAdapterConfig, ParametricMetaAdapterConfig
+from .adapter_modeling import MetaAdapter, Adapter
+from .adapter_utils import HyperNetUpSampler, HyperNetDownSampler
+
+
+class AdapterController(nn.Module):
+  """Implements Adapter controller module which controls the logics of
+  putting adapter layers within transformer's layers."""
+
+  def __init__(self, config):
+    super().__init__()
+    self.config = config
+    self.adapters = nn.ModuleDict(dict())
+    self.tasks = config.tasks
+    self.adapters = self.construct_adapters(self.tasks)
+    self.task_to_adapter = {task: task for task in self.tasks}
+
+  def set_task_to_adapter_map(self, mapping):
+    self.task_to_adapter = mapping
+
+  def get_task(self, task):
+    return self.task_to_adapter[task]
+
+  def construct_adapters(self, tasks):
+    """
+    Constructs adapter layers and adds them to a dictionary for the given
+    tasks.
+    :param tasks: A list of string containing the task names.
+    """
+    for task in tasks:
+      self.adapters[task] = Adapter(self.config)
+    return self.adapters
+
+  def disable_adapters(self, tasks):
+    """
+    Given a list of tasks, it freezes their corresponding adapter layers'
+    parameters.
+    :param tasks: Given list of tasks.
+    """
+    tasks = self.convert_to_list(tasks)
+    for task in tasks:
+      adapter = self.get_adapter(task)
+      for param in adapter.parameters():
+        param.requires_grad = False
+
+  def convert_to_list(self, tasks):
+    if isinstance(tasks, list):
+      return tasks
+    else:
+      return [tasks]
```

Comment:
Can just write this as
if isinstance .... : 
  return tasks
return [tasks]

Reply:
done, thanks